[quant][graphmode][fx] Add support for weight prepack folding

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43728 [quant][graphmode][fx] Add support for weight prepack folding**
* #43581 [quant][graphmode][fx] Add top level APIs
* #43534 [quant][graphmode][fx][fix] enable per channel quantization for functional ops

Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385090](https://our.internmc.facebook.com/intern/diff/D23385090)

jerryzh168 committed:
[quant][graphmode][fx] Add support for weight prepack folding

Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
This is WIP and depends on https://github.com/pytorch/pytorch/pull/43666

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight prepack folding"


Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight prepack folding"


Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385090](https://our.internmc.facebook.com/intern/diff/D23385090)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -52,6 +52,18 @@ def quantize_node(node, activation_post_process):
 def quantize(quantizer, node):
     quantize_node(node, quantizer.activation_post_process_map[node.name])
 
+# returns a function that can get a new attribute name for module with given prefix
```
can we add a real example to the docblock?  It would make the code easier to read, someone can just use it without reading every single line.

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -926,6 +954,87 @@ def get_next_i(module, qparams):
             delattr(observed_root, n)
         return GraphModule(observed_root, self.quantized_graph)
 
+    def _fold_weight(self, quantized):
```
can we add a high level docblock?  I think just adding what the PR has in the summary here would be helpful.

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -577,15 +589,40 @@ def convert(self, quantizer, node, load_arg, debug=False):
+
+# Patterns for weight prepack op folding
+# Base Weight Prepack Folding Pattern Handler
+class FoldingHandler(ABC):
```
makes sense, would this be used in later PRs?  Wasn't sure how it's used here.

jerryzh168 replied:
this will be removed, I thought I need this in the beginning, but turns out we don't

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -926,6 +954,87 @@ def get_next_i(module, qparams):
+                node = frontier.pop()
+                all_args = list(node.args) + list(node.kwargs.values())
+                for arg in all_args:
+                    print('looking at arg:', arg)
```
doesn't have to be for this PR, but should we have some kind of logging levels in the future?  This seems really useful when developing or debugging, and ideally we can keep it off by default so users don't see the low level.

jerryzh168 replied:
I think so, not sure what is the the recommended python logging for pytorch, I remember someone posted before and dima answered just use some global flag to control the logging level?

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight prepack folding"


Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385090](https://our.internmc.facebook.com/intern/diff/D23385090)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight prepack folding"


Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385090](https://our.internmc.facebook.com/intern/diff/D23385090)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight prepack folding"


Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385090](https://our.internmc.facebook.com/intern/diff/D23385090)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight prepack folding"


Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385090](https://our.internmc.facebook.com/intern/diff/D23385090)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight prepack folding"


Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385090](https://our.internmc.facebook.com/intern/diff/D23385090)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight prepack folding"


Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385090](https://our.internmc.facebook.com/intern/diff/D23385090)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight prepack folding"


Summary:
Trace back from the weight node util we hit getattr, reconstruct the graph module with the traced nodes
and run the graph module to pack the weight. then replace the original chain of ops with the packed weight.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385090](https://our.internmc.facebook.com/intern/diff/D23385090)

[ghstack-poisoned]
